src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/Interop/LdapPal.Linux.cs

```diff
@@ -99,7 +99,18 @@ internal static int ParseResultReferral(ConnectionHandle ldapHandle, IntPtr resu
         // This option is not supported in Linux, so it would most likely throw.
         internal static int SetServerCertOption(ConnectionHandle ldapHandle, LdapOption option, VERIFYSERVERCERT outValue) => Interop.Ldap.ldap_set_option_servercert(ldapHandle, option, outValue);
 
-        internal static int BindToDirectory(ConnectionHandle ld, string who, string passwd) => Interop.Ldap.ldap_sasl_bind(ld, who, null, new berval(passwd), IntPtr.Zero, IntPtr.Zero, IntPtr.Zero);
+        internal static int BindToDirectory(ConnectionHandle ld, string who, string passwd)
+        {
+            berval passwordBerval = new berval { bv_len = passwd.Length, bv_val = Marshal.StringToHGlobalAuto(passwd) };
```

Comment:
Similar note to before. this pattern is fine of course:
```c#
            IntPtr ptr = Marshal.AllocHGlobal(...);
            try
            {
                        // use
            }
            finally
            {
                     Marshal.FreeHGlobal(ptr);
           }
```
but when it's not quite that clean it gets a bit more muddy. So I'd be inclined to move the allocation inside the try.

Reply:
@jkotas any preference?

Reply:
> but when it's not quite that clean it gets a bit more muddy. So I'd be inclined to move the allocation inside the try.

Agree.

Also, `berval` and similar types can be `struct`s. There is a lot of unnecessary overhead and allocations caused by them being `classes`.

Reply:
> Also, berval and similar types can be structs. There is a lot of unnecessary overhead and allocations caused by them being classes.

Agreed but I suggest we don't need to grow this PR, it could be later.

Reply:
To clarify, you instead want this?

```csharp
IntPtr ptr = null;
try
{
    IntPtr ptr = Marshal.AllocHGlobal(...);                        
    // use
}
finally
{
    Marshal.FreeHGlobal(ptr);
}
```

I can do that.

> Also, berval and similar types can be structs. There is a lot of unnecessary overhead and allocations caused by them being classes.

For my understanding, if `berval` were a struct, we would just need to add a `ref` keyword for `berval*` arguments since structs are blittable, right? I may be able to follow up with another PR later if that improves performance.

Reply:
For your first question - yes, thanks.